File: android/app/src/main/java/com/alexivanov/snapsell/ui/capture/CaptureScreen.kt
package com.alexivanov.snapsell.ui.capture

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.em
import com.alexivanov.snapsell.AppContainer
import com.alexivanov.snapsell.R
import com.alexivanov.snapsell.ui.common.ErrorText
import com.alexivanov.snapsell.ui.common.LoadingBox
import com.alexivanov.snapsell.ui.common.LucideIcon
import com.alexivanov.snapsell.ui.common.PrimaryButton
import com.alexivanov.snapsell.ui.common.SnapTopBar
import com.alexivanov.snapsell.ui.common.Spinner
import com.alexivanov.snapsell.ui.theme.SnapType
import com.alexivanov.snapsell.ui.theme.snapColors
import com.alexivanov.snapsell.vision.CameraScreen
import com.alexivanov.snapsell.vision.CaptureSession
import com.alexivanov.snapsell.vision.PhotoStore
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/** Full-bleed camera. Nothing else on screen. */
@Composable
fun CaptureScreen(container: AppContainer, onReview: () -> Unit, onBack: () -> Unit) {
    val scope = rememberCoroutineScope()
    val c = snapColors
    var preparing by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }

    if (preparing) {
        Box(Modifier.fillMaxSize().background(c.bg)) { LoadingBox(label = "Preparing photo") }
        return
    }

    Box(Modifier.fillMaxSize().background(Color.Black)) {
        CameraScreen(
            photoStore = container.photoStore,
            onCaptured = { file ->
                preparing = true
                scope.launch {
                    val bitmap = withContext(container.dispatchers.io) {
                        container.photoStore.decodeScaled(file, PhotoStore.WORK_LONG_EDGE)
                    }
                    if (bitmap == null) {
                        preparing = false
                        error = "Could not read the photo. Try again."
                    } else {
                        container.captureSession.current?.workBitmap?.recycle()
                        container.captureSession.current = CaptureSession.Photo(file, bitmap)
                        onReview()
                    }
                }
            },
            onError = { error = it },
            denied = { requestAgain ->
                Column(Modifier.fillMaxSize().background(c.bg)) {
                    SnapTopBar(title = "Capture", onBack = onBack)
                    Column(Modifier.padding(24.dp)) {
                        Text("Camera access needed", style = SnapType.sectionHead, color = c.text)
                        Text("SnapSell photographs your items with the camera. Nothing is uploaded until you confirm what to sell.", style = SnapType.body, color = c.text.copy(alpha = 0.75f), modifier = Modifier.padding(top = 8.dp, bottom = 20.dp))
                        PrimaryButton("Allow camera", onClick = requestAgain, modifier = Modifier.fillMaxWidth())
                    }
                }
            },
        ) { controls ->
            // Four 28dp white corner brackets, 3dp stroke, inset 20dp; the top pair at y = 72dp.
            Canvas(Modifier.fillMaxSize().statusBarsPadding()) {
                val stroke = 3.dp.toPx()
                val len = 28.dp.toPx()
                val inset = 20.dp.toPx()
                val top = 72.dp.toPx()
                val bottom = size.height - inset - 120.dp.toPx()
                val left = inset
                val right = size.width - inset
                val white = Color.White
                fun bracket(x: Float, y: Float, dx: Float, dy: Float) {
                    drawLine(white, Offset(x, y), Offset(x + dx * len, y), stroke, StrokeCap.Square)
                    drawLine(white, Offset(x, y), Offset(x, y + dy * len), stroke, StrokeCap.Square)
                }
                bracket(left, top, 1f, 1f)
                bracket(right, top, -1f, 1f)
                bracket(left, bottom, 1f, -1f)
                bracket(right, bottom, -1f, -1f)
            }

            // Back arrow, white, in a 52dp box.
            Box(
                Modifier
                    .statusBarsPadding()
                    .padding(start = 4.dp, top = 4.dp)
                    .size(52.dp)
                    .clickable(onClick = onBack),
                contentAlignment = Alignment.Center,
            ) {
                LucideIcon(R.drawable.ic_lucide_arrow_left, "Back", size = 24.dp, tint = Color.White)
            }

            Column(Modifier.align(Alignment.BottomCenter).navigationBarsPadding(), horizontalAlignment = Alignment.CenterHorizontally) {
                ErrorText(error, Modifier.padding(bottom = 12.dp))
                Text(
                    "Point at one item, or a pile".uppercase(),
                    style = SnapType.bodySmall.copy(letterSpacing = 0.06.em),
                    color = Color.White.copy(alpha = 0.9f),
                    // Sits just above the shutter, whose top edge is 100dp from the bottom.
                    modifier = Modifier.padding(bottom = 12.dp),
                )
                // Shutter: 76dp circle, 3dp white ring, 6dp gap, white fill, 30dp accent glyph.
                Box(
                    Modifier
                        .padding(bottom = 24.dp)
                        .size(76.dp)
                        .clip(CircleShape)
                        .clickable(enabled = !controls.capturing, onClick = controls.capture),
                    contentAlignment = Alignment.Center,
                ) {
                    Canvas(Modifier.fillMaxSize()) {
                        val ring = 3.dp.toPx()
                        drawCircle(Color.White, radius = size.minDimension / 2 - ring / 2, style = androidx.compose.ui.graphics.drawscope.Stroke(ring))
                        drawCircle(Color.White, radius = size.minDimension / 2 - ring - 6.dp.toPx())
                    }
                    if (controls.capturing) {
                        Spinner(size = 30.dp, color = c.accent)
                    } else {
                        // The shutter fill is white in both themes, so the glyph keeps the light accent.
                        LucideIcon(R.drawable.ic_lucide_camera, "Take photo", size = 30.dp, tint = ShutterGlyph)
                    }
                }
            }
        }
    }
}

private val ShutterGlyph = Color(0xFFEC3013)

File: ios/SnapSell/UI/Screens/Capture/CaptureScreen.swift
import AVFoundation
import PhotosUI
import SwiftUI

/// Full-bleed camera. On devices without one (the simulator) a photo picker
/// stands in so the rest of the flow can be exercised.
struct CaptureScreen: View {
    @Environment(AppContainer.self) private var app
    @Environment(Router.self) private var router
    @Environment(\.snap) private var c
    @State private var trigger = false
    @State private var capturing = false
    @State private var preparing = false
    @State private var error: String?
    @State private var permission = AVCaptureDevice.authorizationStatus(for: .video)
    @State private var pickerItem: PhotosPickerItem?

    private let hasCamera = CameraView.hasCamera

    var body: some View {
        Group {
            if preparing {
                VStack { LoadingBox("Preparing photo").padding(16); Spacer() }.screenBackground()
            } else if hasCamera && permission == .denied || permission == .restricted {
                denied
            } else if hasCamera && permission == .notDetermined {
                Color.black.ignoresSafeArea().task {
                    let ok = await AVCaptureDevice.requestAccess(for: .video)
                    permission = ok ? .authorized : .denied
                }
            } else {
                camera
            }
        }
        .snapScreen()
    }

    private var camera: some View {
        ZStack {
            Color.black.ignoresSafeArea()
            if hasCamera {
                CameraView(trigger: $trigger, onCapture: { image in capturing = false; handle(image) }, onError: { e in capturing = false; error = e })
                    .ignoresSafeArea()
            } else {
                VStack(spacing: 12) {
                    Image(systemName: "camera.metering.unknown").font(.system(size: 40)).foregroundStyle(.white.opacity(0.7))
                    Text("No camera on this device").snap(SnapType.body).foregroundStyle(.white.opacity(0.8))
                    Text("Pick a photo from the library instead.").snap(SnapType.bodySmall).foregroundStyle(.white.opacity(0.6))
                }
            }
            brackets
            VStack {
                HStack {
                    Button { router.pop() } label: {
                        Image(systemName: "arrow.left").font(.system(size: 24, weight: .bold)).foregroundStyle(.white).frame(width: 52, height: 52)
                    }
                    .padding(.leading, 4).padding(.top, 4)
                    Spacer()
                }
                Spacer()
                if let error { ErrorText(error).padding(.bottom, 12) }
                Text("Point at one item, or a pile".uppercased())
                    .font(.custom("Archivo-Regular", size: 13)).kerning(13 * 0.06)
                    .foregroundStyle(.white.opacity(0.9)).padding(.bottom, 12)
                shutter.padding(.bottom, 24)
            }
        }
    }

    private var shutter: some View {
        Group {
            if hasCamera {
                Button {
                    guard !capturing else { return }
                    capturing = true
                    error = nil
                    trigger = true
                } label: { shutterBody }
                .buttonStyle(.plain)
                .disabled(capturing)
                .accessibilityLabel("Take photo")
            } else {
                PhotosPicker(selection: $pickerItem, matching: .images) { shutterBody }
                    .onChange(of: pickerItem) { _, item in
                        guard let item else { return }
                        capturing = true
                        Task {
                            defer { capturing = false; pickerItem = nil }
                            if let data = try? await item.loadTransferable(type: Data.self), let image = UIImage(data: data) {
                                handle(image)
                            } else {
                                error = "Couldn't read that photo."
                            }
                        }
                    }
            }
        }
    }

    private var shutterBody: some View {
        ZStack {
            Circle().stroke(.white, lineWidth: 3)
            Circle().fill(.white).padding(9)
            if capturing {
                ProgressView().tint(Color(hex: 0xEC3013))
            } else {
                Image(systemName: "camera.fill").font(.system(size: 28, weight: .bold)).foregroundStyle(Color(hex: 0xEC3013))
            }
        }
        .frame(width: 76, height: 76)
    }

    /// Four 28pt white corner brackets, 3pt stroke, inset 20pt.
    private var brackets: some View {
        GeometryReader { geo in
            let len: CGFloat = 28, inset: CGFloat = 20
            let top: CGFloat = 72
            let bottom = geo.size.height - inset - 120
            let right = geo.size.width - inset
            Path { p in
                for (x, y, dx, dy) in [(inset, top, 1.0, 1.0), (right, top, -1.0, 1.0), (inset, bottom, 1.0, -1.0), (right, bottom, -1.0, -1.0)] {
                    p.move(to: CGPoint(x: x, y: y)); p.addLine(to: CGPoint(x: x + dx * len, y: y))
                    p.move(to: CGPoint(x: x, y: y)); p.addLine(to: CGPoint(x: x, y: y + dy * len))
                }
            }
            .stroke(.white, style: StrokeStyle(lineWidth: 3, lineCap: .square))
        }
        .allowsHitTesting(false)
    }

    private var denied: some View {
        VStack(alignment: .leading, spacing: 0) {
            SnapTopBar("Capture", onBack: { router.pop() })
            VStack(alignment: .leading, spacing: 0) {
                SectionHead("Camera access needed")
                Text("SnapSell photographs your items with the camera. Nothing is uploaded until you confirm what to sell.")
                    .snap(SnapType.body).foregroundStyle(c.text.opacity(0.75)).padding(.top, 8).padding(.bottom, 20)
                SnapButton("Open Settings") {
                    if let url = URL(string: UIApplication.openSettingsURLString) { UIApplication.shared.open(url) }
                }
            }
            .padding(24)
            Spacer()
        }
        .screenBackground()
    }

    private func handle(_ image: UIImage) {
        preparing = true
        error = nil
        let store = app.photoStore
        Task {
            do {
                let (file, work) = try await Task.detached(priority: .userInitiated) {
                    let file = try store.saveOriginal(image)
                    let work = image.normalized(maxLongEdge: PhotoStore.workLongEdge)
                    return (file, work)
                }.value
                app.captureSession.current = CaptureSession.Photo(file: file, workImage: work)
                router.replaceTop(with: .review)
            } catch {
                preparing = false
                self.error = "Could not save the photo. Try again."
            }
        }
    }
}
